### PR TITLE
[FLINK-24779][table-planner] Introduce numeric casting rules

### PR DESCRIPTION
**Pull request by slinkydeveloper**
## What is the purpose of the change

Introduce numeric casting rules, cleaning up old code.

## Brief change log

* Add new casting rules for primitives/decimal <-> primitives/decimal castings, cleaning up old code in `ScalarOperatorGens` and `CodeGenUtils`
* Make sure `IdentityCastRule` can handle noop interval/int/long castings
* Introduce `ExpressionCodeGeneratorCastRule` to expose `generateExpression`
* Fix box/unbox issue for `CastExecutor` generation

## Verifying this change

This change adds new cases to the `CastRulesTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/ExpressionCodeGeneratorCastRule.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+@Internal
+public interface ExpressionCodeGeneratorCastRule<IN, OUT> extends CodeGeneratorCastRule<IN, OUT> {
+
+    /**
+     * Generate a Java expression performing the casting. This expression can be wrapped in another
+     * expression, or assigned to a variable or returned from a function.
+     *
+     * <p>NOTE: the {@code inputTerm} is always either a primitive or a non-null object.
+     */
+    String generateExpression(
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/ExpressionCodeGeneratorCastRule.java:39]: add JavaDoc for the return type? also either a primitive or a non-null object?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala [hunk 2]
```diff
@@ -195,6 +195,38 @@ object CodeGenUtils {
     case _ => boxedTypeTermForType(t)
   }
 
+  /**
+   * Execute primitive unboxing.
+   */
+  def unbox(term: String, ty: LogicalType): String = ty.getTypeRoot match {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:201]: nit: why in Scala? We could have moved it to CastUtils where also `CastRuleUtils.ternaryOperator` is located?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/rules/AbstractNullAwareCodeGeneratorCastRule.java [hunk 3]
```diff
@@ -59,8 +60,8 @@ public CastCodeBlock generateCodeBlock(
             LogicalType targetType) {
         final CastRuleUtils.CodeWriter writer = new CastRuleUtils.CodeWriter();
 
-        // Result of a casting can be null only and only if the input is null
-        final boolean isResultNullable = inputType.isNullable();
+        // Result of a casting is null if the input is null or the output primitive type is nullable
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/rules/AbstractNullAwareCodeGeneratorCastRule.java:63]: nit: is this comment helpful? it basically describes the actually very readable expression below.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/rules/DecimalToNumericPrimitiveCastRule.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import static org.apache.flink.table.planner.codegen.calls.BuiltInMethods.DECIMAL_TO_INTEGRAL;
+import static org.apache.flink.table.planner.functions.casting.rules.CastRuleUtils.castToPrimitive;
+import static org.apache.flink.table.planner.functions.casting.rules.CastRuleUtils.staticCall;
+
+/**
+ * {@link LogicalTypeRoot#DECIMAL} to {@link LogicalTypeFamily#INTEGER_NUMERIC} and {@link
+ * LogicalTypeFamily#APPROXIMATE_NUMERIC} cast rule.
+ */
+@Internal
+public class DecimalToNumericPrimitiveCastRule
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/rules/DecimalToNumericPrimitiveCastRule.java:41]: can we move the rules one package level up? we do the same for `DataStructureConverter`. This allows to have all rules package private and we don't spam the classpath. And having everything under `casting` should be abstraction enough.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/rules/NumericPrimitiveCastRule.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+public class NumericPrimitiveCastRule
+        extends AbstractExpressionCodeGeneratorCastRule<Number, Number> {
+
+    public static final NumericPrimitiveCastRule INSTANCE = new NumericPrimitiveCastRule();
+
+    private NumericPrimitiveCastRule() {
+        super(CastRulePredicate.builder().predicate(NumericPrimitiveCastRule::matches).build());
+    }
+
+    private static boolean matches(LogicalType x, LogicalType y) {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/rules/NumericPrimitiveCastRule.java:48]: `src` /`dest` or something meaningful?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/rules/NumericPrimitiveToDecimalCastRule.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    @Override
+    public String generateExpression(
+            CodeGeneratorCastRule.Context context,
+            String inputTerm,
+            LogicalType inputLogicalType,
+            LogicalType targetLogicalType) {
+        final DecimalType targetDecimalType = (DecimalType) targetLogicalType;
+        return methodCall(
+                className(DecimalDataUtils.class),
+                "castFrom",
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/rules/NumericPrimitiveToDecimalCastRule.java:63]: add this to `BuiltInMethods`

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/casting/CastRulesTest.java [hunk 7]
```diff
@@ -236,6 +471,35 @@
...
                                 StringData.fromString("2020-11-11T18:08:01.123")),
+                CastTestSpecBuilder.testCastTo(DECIMAL(5, 3))
+                        .fromCase(
+                                DECIMAL(4, 3),
+                                DecimalData.fromBigDecimal(new BigDecimal("9.87"), 4, 3),
+                                DecimalData.fromBigDecimal(new BigDecimal("9.870"), 5, 3))
+                        .fromCase(
+                                TINYINT(),
+                                (byte) -1,
+                                DecimalData.fromBigDecimal(new BigDecimal("-1.000"), 5, 3))
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/casting/CastRulesTest.java:482]: introduce a help method for the very repetitive `DecimalData.fromBigDecimal(new BigDecimal("-1.000"), 5, 3)`?

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the PR @slinkydeveloper. I added some minor comments. Should be good in the next iteration.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/casting/CastRulesTest.java (hunk 7 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/casting/CastRulesTest.java:482]: This number is used only once in our tests :) Perhaps you're referring to another number?

**Comment by slinkydeveloper** [comment]
@twalthr this is ready for another pass now

**Review by twalthr — APPROVED** [review]
LGTM

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/casting/CastRulesTest.java (hunk 7 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/casting/CastRulesTest.java:482]: What you could have done is simply `decimal("-1.000")` or `decimal("-1.000", 5, 3)` and abstract away the verbose `DecimalData.fromBigDecimal(new BigDecimal("-1.000"), 5, 3)` for readbility.

**Pull request closed without merging**